File: forms/neurodiversity-adjustment-request/front-end-with-svelte/src/lib/engine/completeness-rules.ts
import type { NeurodiversityAdjustmentRequest, FiredRule } from './types';
import { anyCondition, anyDifficulty, anyAdjustment } from './utils';

/** A mandatory request field, used to compute Axis C completeness. */
interface FieldCheck {
	ruleId: string;
	label: string;
	weight: number;
	present: (r: NeurodiversityAdjustmentRequest) => boolean;
}

/**
 * The mandatory request fields, weighted by importance to the request. The
 * neurodivergent profile, functional difficulties, and requested adjustments
 * are weighted highest. Completeness is the weighted percentage of fields
 * supplied.
 */
const fields: FieldCheck[] = [
	{
		ruleId: 'R-COMPLETE-CONDITIONS',
		label: 'neurodivergent profile',
		weight: 3,
		present: (r) => anyCondition(r)
	},
	{
		ruleId: 'R-COMPLETE-DIFFICULTIES',
		label: 'functional difficulties',
		weight: 3,
		present: (r) => anyDifficulty(r)
	},
	{
		ruleId: 'R-COMPLETE-ADJUSTMENTS',
		label: 'requested adjustments',
		weight: 3,
		present: (r) => anyAdjustment(r)
	},
	{
		ruleId: 'R-COMPLETE-TASKS',
		label: 'tasks and situations affected',
		weight: 2,
		present: (r) => r.tasksSituationsAffected.trim() !== ''
	},
	{
		ruleId: 'R-COMPLETE-CONSENT',
		label: 'disclosure consent',
		weight: 2,
		present: (r) => r.disclosureConsent === true
	},
	{
		ruleId: 'R-COMPLETE-WORKER-NAME',
		label: 'worker name',
		weight: 1,
		present: (r) => r.workerName.trim() !== ''
	},
	{
		ruleId: 'R-COMPLETE-JOB-TITLE',
		label: 'job title',
		weight: 1,
		present: (r) => r.workerJobTitle.trim() !== ''
	},
	{
		ruleId: 'R-COMPLETE-MANAGER',
		label: 'manager / HR contact',
		weight: 1,
		present: (r) => r.managerName.trim() !== ''
	},
	{
		ruleId: 'R-COMPLETE-REQUEST-DATE',
		label: 'request date',
		weight: 1,
		present: (r) => r.requestDate.trim() !== ''
	},
	{
		ruleId: 'R-COMPLETE-ADJUSTMENTS-DETAIL',
		label: 'requested-adjustments detail',
		weight: 1,
		present: (r) => r.adjustmentsRequestedDetail.trim() !== ''
	}
];

/**
 * Axis C — request completeness.
 *
 * Returns the weighted percentage (0–100, rounded) of mandatory request fields
 * that are present, plus an audit-trail rule for each missing field.
 */
export function gradeCompleteness(r: NeurodiversityAdjustmentRequest): {
	completenessPercent: number;
	firedRules: FiredRule[];
} {
	const firedRules: FiredRule[] = [];
	const totalWeight = fields.reduce((sum, f) => sum + f.weight, 0);
	let presentWeight = 0;

	for (const field of fields) {
		if (field.present(r)) {
			presentWeight += field.weight;
		} else {
			firedRules.push({
				ruleId: field.ruleId,
				axis: 'completeness',
				category: 'missing-field',
				description: `Mandatory field missing: ${field.label}.`
			});
		}
	}

	const completenessPercent = Math.round((presentWeight / totalWeight) * 100);
	return { completenessPercent, firedRules };
}

File: forms/neurodiversity-adjustment-request/back-end-with-loco/src/engine/completeness_rules.rs
//! Axis C — request completeness.
//!
//! Ported from `front-end-with-svelte/src/lib/engine/completeness-rules.ts`.
//! Completeness is the weighted percentage (0–100, rounded) of mandatory
//! request fields that are present; each missing field emits an audit-trail rule
//! on the `completeness` axis with category `missing-field`.

use super::types::{FiredRule, NeurodiversityAdjustmentRequest};
use super::utils::{any_adjustment, any_condition, any_difficulty};

/// A mandatory request field, used to compute Axis C completeness.
struct FieldCheck {
    rule_id: &'static str,
    label: &'static str,
    weight: i32,
    present: fn(&NeurodiversityAdjustmentRequest) -> bool,
}

/// The mandatory request fields, with the neurodivergent profile, functional
/// difficulties, and requested adjustments weighted highest (total weight = 18).
const FIELDS: &[FieldCheck] = &[
    FieldCheck {
        rule_id: "R-COMPLETE-CONDITIONS",
        label: "neurodivergent profile",
        weight: 3,
        present: any_condition,
    },
    FieldCheck {
        rule_id: "R-COMPLETE-DIFFICULTIES",
        label: "functional difficulties",
        weight: 3,
        present: any_difficulty,
    },
    FieldCheck {
        rule_id: "R-COMPLETE-ADJUSTMENTS",
        label: "requested adjustments",
        weight: 3,
        present: any_adjustment,
    },
    FieldCheck {
        rule_id: "R-COMPLETE-TASKS",
        label: "tasks and situations affected",
        weight: 2,
        present: |r| !r.tasks_situations_affected.trim().is_empty(),
    },
    FieldCheck {
        rule_id: "R-COMPLETE-CONSENT",
        label: "disclosure consent",
        weight: 2,
        present: |r| r.disclosure_consent,
    },
    FieldCheck {
        rule_id: "R-COMPLETE-WORKER-NAME",
        label: "worker name",
        weight: 1,
        present: |r| !r.worker_name.trim().is_empty(),
    },
    FieldCheck {
        rule_id: "R-COMPLETE-JOB-TITLE",
        label: "job title",
        weight: 1,
        present: |r| !r.worker_job_title.trim().is_empty(),
    },
    FieldCheck {
        rule_id: "R-COMPLETE-MANAGER",
        label: "manager / HR contact",
        weight: 1,
        present: |r| !r.manager_name.trim().is_empty(),
    },
    FieldCheck {
        rule_id: "R-COMPLETE-REQUEST-DATE",
        label: "request date",
        weight: 1,
        present: |r| !r.request_date.trim().is_empty(),
    },
    FieldCheck {
        rule_id: "R-COMPLETE-ADJUSTMENTS-DETAIL",
        label: "requested-adjustments detail",
        weight: 1,
        present: |r| !r.adjustments_requested_detail.trim().is_empty(),
    },
];

/// The result of grading axis C.
pub struct Completeness {
    /// The weighted completeness percentage (0–100, rounded).
    pub completeness_percent: i32,
    /// One audit-trail rule for each missing field.
    pub fired_rules: Vec<FiredRule>,
}

/// Grade axis C — request completeness as the weighted percentage of mandatory
/// request fields present, with an audit-trail rule for each missing field.
#[must_use]
pub fn grade_completeness(r: &NeurodiversityAdjustmentRequest) -> Completeness {
    let mut fired_rules: Vec<FiredRule> = Vec::new();
    let total_weight: i32 = FIELDS.iter().map(|f| f.weight).sum();
    let mut present_weight: i32 = 0;

    for field in FIELDS {
        if (field.present)(r) {
            present_weight += field.weight;
        } else {
            fired_rules.push(FiredRule::new(
                field.rule_id,
                "completeness",
                "missing-field",
                &format!("Mandatory field missing: {}.", field.label),
            ));
        }
    }

    // Round-half-up to match JavaScript's Math.round on non-negative values.
    let completeness_percent =
        ((f64::from(present_weight) / f64::from(total_weight)) * 100.0).round() as i32;

    Completeness {
        completeness_percent,
        fired_rules,
    }
}
